docs/design/features/greenthreads.md

```diff
@@ -0,0 +1,233 @@
+### Ways in which green threads are a leaky abstraction
+There are a number of details about the green thread design with are “leaky” abstractions
+-	If a green thread calls into native code and then back into managed, we cannot yield the thread, and any attempts will simply fall back to a standard thread blocking wait
+-	Interaction with native code which has thread static state is risky. If a bit of interop exposes native thread state to managed code, and then the thread changes out underneath that code interesting behaviors may occur
+
+### Example of change to plaintext demo app
+
+Before
+```csharp
+var response = context.Response;
+response.StatusCode = 200;
+response.ContentType = "text/plain";
+response.ContentLength = payload.Length;
+return response.Body.WriteAsync(payload).AsTask();
+```
+After
+```csharp
+var response = context.Response;
+response.StatusCode = 200;
+response.ContentType = "text/plain";
+response.ContentLength = payload.Length;
+// This is async IO under the covers!
+response.Body.Write(payload);
+```
+
+## Measured results
+A number of results were gathered to validate the performance characteristics of the system. Mostly they indicated that there were some quality of implementation issues, but that implementation of green threads was a feasible approach.
+### Socket Microbenchmarks
+The microbenchmarks measure some overheads and some simple in-proc socket operations for comparison.
+
+Yielding a green thread and resuming it combined takes about 500 ns. By comparison, two context switches of a regular thread take about 8 µs. This may be interesting for reducing the CPU overhead of spin-waiting, and custom thread scheduling. Awaiting an async operation and continuing it takes only 110 ns, so there is a bit of a gap. However, async/await can become viral and longer await chains would each incur the overhead whereas green thread yield/resume would not.
+
+![AwaitVsWait](greenthreads_diagrams/AwaitVsWait.png)
+
+- [Deep await on regular thread VS Deep await on green thread] Awaits have more overhead currently when they happen on green threads. Since the continuation need to be marshaled to run on a green thread, they miss the fast path, involving more allocation, and some more overhead in switching to a green thread. There may be ways to reduce this overhead, but there will probably be some overhead that is a necessary consequence of enabling both green and non-green threads to interoperate. For instance, if thread pool worker threads ran as green threads by default, some of this overhead would likely go away.
```

Comment:
```suggestion
- **Deep await on regular thread vs Deep await on green thread** -- Awaits have more overhead currently when they happen on green threads. Since the continuation need to be marshaled to run on a green thread, they miss the fast path, involving more allocation, and some more overhead in switching to a green thread. There may be ways to reduce this overhead, but there will probably be some overhead that is a necessary consequence of enabling both green and non-green threads to interoperate. For instance, if thread pool worker threads ran as green threads by default, some of this overhead would likely go away.
```